onnxruntime/core/providers/cpu/tensor/utils.h

```diff
@@ -139,11 +139,27 @@ struct SliceSkips : std::vector<int64_t> {
 // This provides easy sequential iteration over a subset of a tensor given a span of starts & extents
 template <typename T>
 struct SliceIterator {
-  SliceIterator(const Tensor& tensor, gsl::span<const int64_t> starts, gsl::span<const int64_t> extents)
-      : tensor_(tensor), extents_(extents), skips_(tensor, extents), indices_(extents.size(), 0) {
+    SliceIterator(const Tensor& tensor, gsl::span<const int64_t> starts, gsl::span<const int64_t> extents)
+        : tensor_(tensor), extents_(extents), skips_(tensor_.Shape(), extents), indices_(extents.size(), 0) {
     auto& dims = tensor_.Shape().GetDims();
     ORT_ENFORCE(static_cast<ptrdiff_t>(dims.size()) == starts.size() && static_cast<ptrdiff_t>(dims.size()) == extents.size());
 
+    Init(dims, starts);
+  }
+    
+    // This construct takes a explicit tensor_shape which might be different from the shape defined in input tensor.
+    // The explicit tensor_shape usually has inner most axis flattened. For example, given shape[1,4,4,2], if last axis
+    // does not have padding or slice, then it will be flattened as [1,4,8] for better performance (One inner most copy instead of 4).
+    SliceIterator(const Tensor& tensor, const TensorShape& tensor_shape, gsl::span<const int64_t> starts, gsl::span<const int64_t> extents)
+      : tensor_(tensor), extents_(extents), skips_(tensor_shape, extents), indices_(extents.size(), 0) {
+    auto& dims = tensor_shape.GetDims();
+    ORT_ENFORCE(static_cast<ptrdiff_t>(dims.size()) == starts.size() && static_cast<ptrdiff_t>(dims.size()) == extents.size());
```

Comment:
Maybe move this (and the same one above) into the Init() function and change it to use extents_.size() (since that's what's available inside Init())?

Reply:
fixed.